theano/tensor/nnet/abstract_conv2d.py

```diff
@@ -0,0 +1,686 @@
+                                     broadcastable=broadcastable)()
+        output = kern.type.clone(broadcastable=broadcastable)()
+        return Apply(self, [kern, topgrad, shape], [output])
+
+
+    def perform(self, node, nodename, inp, out_):
+        raise NotImplementedError('AbstractConv2d_gradWeight theano optimization failed')
+
+    def grad(self, inp, grads):
+        weights, top = inp[:2]
+        bottom, = grads
+        d_weights = AbstractConv2d_gradWeights(self.imshp, self.kshp,
+                                               self.bsize,
+                                               self.border_mode,
+                                               self.subsample)(bottom, top, weights.shape[-2:])
+        d_top = AbstractConv2d(self.imshp, self.filter_shape, self.bsize,
+                               self.border_mode, self.subsample)(bottom, weights)
+        d_height_width = (theano.gradient.DisconnectedType()(),) * 2
+        return (d_weights, d_top) + d_height_width
+
+    def connection_pattern(self, node):
+        return [[1], [1], [0], [0]]  # no connection to height, width
+
+
+### Optimizations should be move in their appropriate files
+
+### move to Gpu optimization
+### Do not replace the AbstractOpt only the inputs
+### Abstract Ops is replaced layer by device_specialized opt
+@local_optimizer([gpu_from_host, AbstractConv2d,
+                  AbstractConv2d_gradWeights,
+                  AbstractConv2d_gradInputs])
+def local_conv2d_gpu_conv(node):
+    """
+    gpu_from_host(AbstractConv) -> AbstractConv(gpu_from_host)
+
+    AbstractConv(host_from_gpu) -> host_from_gpu(AbstractConv)
+    """
+    if isinstance(node.op, GpuFromHost):
+        #gpu_from_host(conv) -> gpu_conv(gpu_from_host)
+        host_input = node.inputs[0]
+        if host_input.owner and \
+                (isinstance(host_input.owner.op, AbstractConv2d) or
+                 isinstance(host_input.owner.op, AbstractConv2d_gradWeights) or
+                 isinstance(host_input.owner.op, AbstractConv2d_gradInputs)):
+
+            conv = host_input.owner.op
+            inps = list(host_input.owner.inputs)
+            inps[0] = gpu_from_host(inps[0])
+            inps[1] = gpu_from_host(inps[1])
+            out = conv(*inps)
+            out = theano.tensor.patternbroadcast(gpu_from_host(out),
+                                                 node.outputs[0].broadcastable)
+            out.values_eq_approx = values_eq_approx_high_tol
+            return [out]
+
+    if (isinstance(node.op, AbstractConv2d) or
+        isinstance(node.op, AbstractConv2d_gradWeights) or
+        isinstance(node.op, AbstractConv2d_gradInputs)):
+        #conv(host_from_gpu) -> host_from_gpu(gpu_conv)
+        inp1 = node.inputs[0]
+        inp2 = node.inputs[1]
+        inp1_on_gpu = (inp1.owner and isinstance(inp1.owner.op, HostFromGpu))
+        inp2_on_gpu = (inp2.owner and isinstance(inp2.owner.op, HostFromGpu))
```

Comment:
Also check if inp1 or inp2 is already directly on the GPU (a CudaNdarrayVariable)